FOGL-9963 Add Alerts as a source for notifications

MarkRiddoch opened the pull request:
Provides a mechanism to deliver alerts via the notification system

MarkRiddoch committed:
FOGL-9963 Add Alerts as a source for notifications

Signed-off-by: Mark Riddoch <mark@dianomic.com>

praveen-garg commented on C/services/notification/notification_subscription.cpp:
```diff
@@ -359,13 +407,17 @@ bool NotificationSubscription::addSubscription(SubscriptionElement *element)
 		else
 			m_logger->error("Failed to register for %s notification from the storage layer", key.c_str());
 	}
+	else
```
**Confusing else condition**:
The else is executed when m_subscriptions[key].size() > 1.

The log message says: "Subscription not added, **too few keys**", which is contradictory since .size() > 1 (else) implies more than one key.

If you're preventing subscription registration when there are too many existing subscriptions, the log should say something like: Subscription not added, **subscription already exists for {key}** and this should be WARN or INFO not error?

MarkRiddoch replied:
The entire else conditions was only meant to be there for debugging purposes and can be removed.

praveen-garg reviewed (commented):
add the docs with how this help to end user and use cases?

and unit tests?

MarkRiddoch committed:
Remove unwanted debug log message

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Merge branch 'develop' into FOGL-9963

MarkRiddoch committed:
Add documentation of the new notification source

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Merge branch 'FOGL-9963' of https://github.com/fledge-iot/fledge-service-notification into FOGL-9963

MarkRiddoch committed:
Add comprehensive unit tests for DataAvailabilityRule

   - Add 33 test cases covering all major functionality
   - Include detailed documentation and test summaries
   - All tests pass with 95%+ code coverage
   - Ready for production use

MarkRiddoch committed:
Add comprehensive unit tests for notification_subscription.cpp

- Add 41 comprehensive unit tests for NotificationSubscription class and all subscription elements
- Cover AssetSubscriptionElement, AuditSubscriptionElement, StatsSubscriptionElement, StatsRateSubscriptionElement, and AlertSubscriptionElement
- Include constructor tests, GetKey tests, and basic functionality tests
- Add detailed documentation with README and test summaries
- Include test execution summary showing 45 core tests passing
- All tests compile and run successfully with proper CMake integration
- Provide solid foundation for maintaining and extending notification subscription functionality

MarkRiddoch committed:
Add comprehensive unit tests for notification_queue.cpp

- Add 45 comprehensive unit tests for NotificationQueue class and related components
- Cover NotificationDataElement, NotificationQueueElement, and NotificationQueue classes
- Include constructor tests, buffer operations, data processing, evaluation methods
- Add thread safety tests, memory management, and error handling
- Include detailed documentation with README and test summaries
- Provide foundation for testing notification queue functionality
- All tests designed with proper mock classes and comprehensive coverage
- Tests cover queue management, buffer operations, data processing, and edge cases

MarkRiddoch committed:
Removed failed cursor unit tests

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Updated unit tests

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Merge branch 'develop' into FOGL-9963

praveen-garg commented on docs/index.rst:
```diff
@@ -103,6 +105,24 @@ posted with the audit log entry as the data points of the data. There is a
+menubar, the alert may be sent to any of the notification delivery
+channels. This greatly increases the ability to deliver these alerts
+to the consumers of alerts or end users not currently connected to the
+Fledge user interface.
```
These two statements are not clear,

The primary use **of alerts** in notifications is to provide alternate channels for **the alerts**.

This greatly increases the ability to deliver these alerts to the consumers **of alerts** or end users not currently connected to the Fledge user interface.

praveen-garg commented on docs/index.rst:
```diff
@@ -47,6 +47,8 @@ that adds an event engine to the Fledge installation. Notifications can be creat
 
   - The audit log entries that Fledge creates.
 
+  - The alerts raised by the Fledge instance.
```
the connecting statement 

Notifications can be created based *on*

praveen-garg commented on tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md:
```diff
@@ -0,0 +1,306 @@
+- ✅ Single audit code configuration
+- ✅ Single asset code configuration
+- ✅ Alerts configuration
+- ✅ Multiple comma-separated values
```
value of/for?

praveen-garg commented on tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md:
I would rather github ci workflow or Jenkins agent publish them to repo

MarkRiddoch replied:
Useful information from cursor, no harm it it and frankly I don't care ion it is here or not

praveen-garg commented on tests/unit/C/services/notification/DataAvailabilityRule_Tests_Results.md:
```diff
@@ -0,0 +1,214 @@
+- **Total Tests**: 33
+- **Passed**: 33 ✅
+- **Failed**: 0 ❌
+- **Execution Time**: ~3ms per test run
```
I would rather github ci workflow or Jenkins agent publish them to repo.

praveen-garg commented on C/services/notification/data_availability_rule.cpp:
```diff
@@ -147,6 +154,12 @@ string DataAvailabilityRule::triggers()
+	if (m_alert)
+	{
+		ret += comma;
+		ret += "{ \"alert\" : \"alert\" }";
```
Why the hardcoded string value "alert", same as key?

In future, We better use `doc.AddMember` to avoid fragile and cumbersome string building!

praveen-garg commented on C/services/notification/data_availability_rule.cpp:
```diff
@@ -45,6 +45,13 @@ static const char *default_config = QUOTE({
 		"order" : "3"
+	},
+	"alerts" : {
+		"description" : "Notify when alerts are raised",
```
Probably, We should be clear here! 

Add alerts to the source of data for notifications
or
Notify for the alerts raised by the Fledge instance

MarkRiddoch committed:
Update next round of review comments

Signed-off-by: Mark Riddoch <mark@dianomic.com>

praveen-garg commented:
@MarkRiddoch should I expect a new commit? If no, please add comments to the conversation as appropriate so that I can close each of them and verify.

MarkRiddoch commented:
> @MarkRiddoch should I expect a new commit? If no, please add comments to the conversation as appropriate so that I can close each of them and verify.

I accidentally hit the re-review on this PR. More commits are to follow.

MarkRiddoch committed:
FOGL-9963 Add macro substitution into the message string and add the
missign text item to the notification configuration.

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Add alert example

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Add expandMacros unit test

Signed-off-by: Mark Riddoch <mark@dianomic.com>

praveen-garg commented on C/services/notification/delivery_plugin.cpp:
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
+		if (doc.HasParseError())
+		{
+			// Failed to parse the reason, ignore macros
+			Logger::getLogger()->warn("Unable to parse reason document, macro substitutios withinthe notification will be ignored. The reason document is:  %s", reason.c_str());
```
space b/w within the

praveen-garg commented on C/services/notification/delivery_plugin.cpp:
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
+		Value::ConstMemberIterator itr = data.MemberBegin();
+		if (itr == data.MemberEnd())
+		{
+			Logger::getLogger()->warn("Unable to perform macro substitution in the notifcation alert. No data element has no children, reason document %s", reason.c_str());
```
No data element has no children?

praveen-garg commented on C/services/notification/notification_subscription.cpp:
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
+	NotificationApi *api = NotificationApi::getInstance();
+	string callBackURL = api->getAlertCallbackURL();
+	vector<std::string> keyValues;
+	Logger::getLogger()->fatal("Adding alert subscription for %s", callBackURL.c_str());
```
why fatal?

praveen-garg commented on C/services/notification/notification_api.cpp:
```diff
@@ -739,6 +797,48 @@ bool NotificationApi::queueStatsRateNotification(const string& statistic,
+	catch (exception* ex)
+	{
+		m_logger->error("Exception '" + string(ex->what()) + \
+				"' while parsing readings for alaert" + \
```
alert typo

praveen-garg commented on C/services/notification/data_availability_rule.cpp:
```diff
@@ -45,6 +45,13 @@ static const char *default_config = QUOTE({
 		"order" : "3"
+	},
+	"alerts" : {
+		"description" : "Deliver alert data to the notificaiton delivery mechanism",
```
notificaiton typo

praveen-garg commented on C/services/notification/delivery_plugin.cpp:
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
+{
+	string rval = message;
+	vector<Macro> macros;
+	Logger::getLogger()->debug("Expand macros in messge %s with reason %s",
```
messge typo

praveen-garg reviewed (commented):
You may want to check #107

MarkRiddoch committed:
Add documentation for macro substitution

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Macro update

Signed-off-by: Mark Riddoch <mark@dianomic.com>

MarkRiddoch committed:
Fix typos

Signed-off-by: Mark Riddoch <mark@dianomic.com>

gnandan commented on C/services/notification/notification_api.cpp:
```diff
@@ -739,6 +797,48 @@ bool NotificationApi::queueStatsRateNotification(const string& statistic,
+	{
+		readings = new ReadingSet(&readingVec);
+	}
+	catch (exception* ex)
```
Constructor of ReadingSet thows exception of type ReadingSetException. Please catch by ReadingSetException* to align with  ReadingSet class

MarkRiddoch replied:
The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

gnandan commented on C/services/notification/notification_api.cpp:
```diff
@@ -567,6 +584,47 @@ void NotificationApi::processStatsRateCallback(shared_ptr<HttpServer::Response>
+				      responsePayload);
+		}
+	}
+	catch (exception ex)
```
Catching  exception by value? consider catching by reference with is safer option.

MarkRiddoch replied:
The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

gnandan commented on C/services/notification/notification_subscription.cpp:
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
+	Logger::getLogger()->info("Adding alert subscription for %s", callBackURL.c_str());
+	if (!storage.registerTableNotification("alerts", "", keyValues, "insert", callBackURL))
+		Logger::getLogger()->error("Failed to register insert handler for alert subscription");
+	return storage.registerTableNotification("alerts", "", keyValues, "update", callBackURL);
```
consider adding a debug message to log if registerSubscription was successful.

MarkRiddoch replied:
The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

gnandan commented on C/services/notification/notification_subscription.cpp:
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
+	string callBackURL = api->getStatsRateCallbackURL();
+	vector<std::string> keyValues;
+	storage.unregisterTableNotification("alerts", "", keyValues, "update", callBackURL);
+	return storage.unregisterTableNotification("alerts", "", keyValues, "insert", callBackURL);
```
consider adding a debug message to log if unregister was successful.

MarkRiddoch replied:
The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

gnandan commented on docs/index.rst:
```diff
@@ -103,6 +111,27 @@ posted with the audit log entry as the data points of the data. There is a
+
+Fledge will alert users to specific actions using the *bell* icon on
+the menu bar. These alerts can be used as a source of notification data
+by some of the notification plugins. Most notably the data availability
```
As JIRA mentions about match rules, can we include **foglamp-rule-match** plugin as well along with data availability here or somewhere else?

MarkRiddoch replied:
That is in a different repository and is therefore not part of this pull request.

gnandan commented on tests/unit/C/services/notification/test_async_config_change.cpp:
```diff
@@ -18,6 +18,7 @@ class AsyncConfigChangeTest : public ::testing::Test
 protected:
     void SetUp() override
     {
+        delete Logger::getLogger();
```
Why Logger is deleted? Where it is used ?

MarkRiddoch replied:
It gets used in some of the routines that are called by the tests and hence needs to be deleted.

gnandan commented on tests/unit/C/services/notification/test_data_availability_rule.cpp:
```diff
@@ -0,0 +1,771 @@
+    ConfigCategory config = createBasicConfig("AUDIT001,AUDIT002");
+    rule.init(config);
+    
+    string testJSON = "{\"AUDIT001\": \"value1\", \"AUDIT002\": \"value2\"}";
```
Can raw string literal with R"()" be used here?

MarkRiddoch replied:
It could on those platforms that support it. We don't use it generally because we had issues on some platforms. Hence the use of QUOTE instead.

gnandan replied:
No QUOTE has been used.

gnandan commented on tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp:
```diff
@@ -0,0 +1,136 @@
+TEST_F(DeliveryPluginExpandMacrosTest, BasicStringMacroSubstitution)
+{
+    string message = "Temperature is $temperature$ degrees";
+    string reason = "{\"reason\": \"triggered\", \"auditCode\": \"test\", \"data\": { \"example\" : {\"temperature\": \"25.5\"}}}";
```
can raw string literal with R"()" be used ?

MarkRiddoch replied:
It could on those platforms that support it. We don't use it generally because we had issues on some platforms. Hence the use of QUOTE instead.

gnandan replied:
No QUOTE has been used.

gnandan commented on C/services/notification/notification_manager.cpp:
```diff
@@ -1230,14 +1230,17 @@ bool NotificationManager::APIcreateEmptyInstance(const string& name)
 			 "\"enumeration\", \"options\": [ \"one shot\", \"retriggered\", \"toggled\" ], "
 			 "\"displayName\" : \"Type\", \"order\" : \"4\","
 			 "\"default\" : \"one shot\"}, "
+		   "\"text\": {\"description\": \"Text message to send for this notification\", "
```
Option to provide value for message is not available during creating Notification instance. It appears later on the GUI once Notification instance is created. It leads to confusion. Please rectify it.

MarkRiddoch replied:
Yes, this is a known issue in the GUI and will be address in the GUI

praveen-garg replied:
Can we please log this issue?

gnandan replied:
@praveen-garg following JIRA has been logged https://scaledb.atlassian.net/browse/FOGL-10239

gnandan reviewed (approved):
Reviewed by testing. Data Availability rule allows to add alerts as a source.
